src/ws/messageTypes.ts

```diff
@@ -0,0 +1,146 @@
+  /** Used to correlate responses to requests. Holds message_id value of the message being replied to. */
+  response_to_id: string;
+}
+
+/** Construct and return either a MessageHeaders or ReplyMessageHeaders given the message type and possible response_to_id value. */
+export function newMessageHeaders<T extends MessageType>(
+  message_type: T,
+  response_to_id?: string,
+): MessageHeaderMap[T] {
+  return {
+    message_type,
+    originator: process.pid.toString(),
+    message_id: randomUUID().toString(),
+    ...(response_to_id ? { response_to_id } : {}),
+  } as MessageHeaderMap[T];
+}
+
+/**
+ * Websocket message structure. Generic over the body payload whose structure
+ * is determined by the message type.
+ **/
+export interface Message<T extends MessageType> {
+  headers: MessageHeaderMap[T];
+  body: MessageBodyMap[T];
+}
+
+/** A message whose headers carry field "response_to_id", indicating is a response message. */
+export interface ResponseMessage<T extends MessageType> extends Message<T> {
+  headers: ReplyMessageHeaders;
+}
+
+/**
+ * Workspace -> sidecar message body, sent when a workspace connects to sidecar,
+ * corresponding to message_type {@link MessageType.WORKSPACE_HELLO}
+ */
+export interface WorkspaceHelloBody {
+  workspace_id: number;
+}
+
+/**
+ * Sidecar -> workspaces message body, sent whenever the total number of authorized websocket connections to sidecar changes.
+ * Corresponds to message_type {@link MessageType.WORKSPACE_COUNT_CHANGED}
+ */
+export interface WorkspacesChangedBody {
+  current_workspace_count: number;
+}
+
+/** Sidecar -> workspace message send if/when sidecar has detected an issue with what extension has said. Will be
+ * sent to the workspace that sent the message that caused the issue, then the websocket disconnected.
+ */
+export interface ProtocolErrorBody {
+  error: string;
+}
+
+/** Type mapping of message type -> corresponding message body type */
+export type MessageBodyMap = {
+  [MessageType.WORKSPACE_HELLO]: WorkspaceHelloBody;
+  [MessageType.WORKSPACE_COUNT_CHANGED]: WorkspacesChangedBody;
+  [MessageType.PROTOCOL_ERROR]: ProtocolErrorBody;
+};
+
+/**
+ * Type mapping of message type -> corresponding headers type.
+ * Dictates which messages whose headers should have `response_to_id` field.`
+ */
+type MessageHeaderMap = {
+  [MessageType.WORKSPACE_HELLO]: MessageHeaders;
+  [MessageType.WORKSPACE_COUNT_CHANGED]: MessageHeaders;
+  [MessageType.PROTOCOL_ERROR]: ReplyMessageHeaders;
+};
+
+/**
+ *  Validate the fields of a message body having just come from JSON with its corresponding message type.
+ *  Alas, TypeScript doesn't have a way to enforce that the body is of the correct type for the message type,
+ *  so we have to do it manually.
+ */
+export function validateMessageBody(messageType: MessageType, body: unknown): void {
+  if (messageType === MessageType.WORKSPACE_COUNT_CHANGED) {
+    if (
+      typeof body !== "object" ||
+      !body ||
```

Comment:
As in `body == null`?

Reply:
Well, and also if body === undefined, as in receiving a really degenerate message like:
```
{
   headers: {....}
}
```

Shuffling to get this condiftion as the first clause, writing a unit test over this. Glaringly missing!

Reply:
New tests written in https://github.com/confluentinc/vscode/pull/754/commits/27555a318536fb09c150221937a70ff0bae6815b .

Reply:
If body is undefined, `== null` still true. That's the only case where reference insensitive `==` comparison is quite useful in JS.

Reply:
Respelled to `== null` by request.